File: Ruby/034.rb
MAX = 999_999
FACTORIALS = [1, 1, 2, 6, 24, 120, 720, 5040, 40320, 362880]

def fact_sum(num)
  num.to_s.chars.map(&:to_i).sum{ |n| FACTORIALS[n] }
end

ans = 0
3.upto(MAX).each do |num|
  if fact_sum(num) == num
    ans += num 
  end
end
puts ans

File: Cpp/034.cpp
#include<iostream>
#define M 2540160
using namespace std;

int arr[]={1,1,2,6,24,120,720,5040,40320,362880};

bool sum(int a)
{
    int ans=0,b,c=a;
    while(a)
    {
        b=a%10;
        ans+=arr[b];
        a/=10;
    }
    if(c==ans) return true;
    else return false;
}

int main()
{
    int ans=0;
    for(int i=5;i<M;i++)
    {
        if(sum(i)) ans+=i;
    }
    cout<<ans;
}
